python/ray/serve/_private/router.py

```diff
@@ -415,7 +425,7 @@ def curr_num_scheduling_tasks(self) -> int:
     @property
     def max_num_scheduling_tasks(self) -> int:
         """Max number of scheduling tasks to run at any time."""
-        return 2 * len(self._replicas)
+        return min(self.max_num_scheduling_tasks_hard_limit, 2 * len(self._replicas))
```

Comment:
For my own understanding: where's 2 * replicas heuristic is coming from?

Reply:
From my brain cells and playing around with some benchmarks :)

Basic thought is to at least achieve pipeline parallelism to each replica

We should definitely keep iterating on these params